From c507742c6bcfbdf7e33f3bc390934f1a1280e655 Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Wed, 8 Oct 2025 12:28:36 +0200
Subject: [PATCH 103/150] hack: soc: qcom: pmic_glink: Avoid platform/rpmsg
 probe race condition

If the rpmsg device probes before the platform device, PMIC GLINK setup
will fail with

6800000.remoteproc:glink-edge.PMIC_RTR_ADSP_APPS.-1.-1: error -ENODEV: no
pmic_glink device to attach to

As a workaround, register the rpmsg driver only after the PMIC GLINK
platform device was probed. Returning -EPROBE_DEFER from the rpmsg driver
does not seem to work because the rpmsg core doesn't handle EPROBE_DEFER(?)

6800000.remoteproc:glink-edge.PMIC_RTR_ADSP_APPS.-1.-1: rpmsg_dev_probe:
failed: -517

This needs more investigation.

Signed-off-by: Abel Vesa <abel.vesa@oss.qualcomm.com>
---
 drivers/soc/qcom/pmic_glink.c | 38 ++++++++++++-----------------------
 1 file changed, 13 insertions(+), 25 deletions(-)

diff --git a/drivers/soc/qcom/pmic_glink.c b/drivers/soc/qcom/pmic_glink.c
index c0a4be5df..cf23c3c11 100644
--- a/drivers/soc/qcom/pmic_glink.c
+++ b/drivers/soc/qcom/pmic_glink.c
@@ -342,8 +342,18 @@ static int pmic_glink_probe(struct platform_device *pdev)
 	__pmic_glink = pg;
 	mutex_unlock(&__pmic_glink_lock);
 
+	ret = register_rpmsg_driver(&pmic_glink_rpmsg_driver);
+	if (ret) {
+		ret = dev_err_probe(&pdev->dev, ret, "failed to register rpmsg driver\n");
+		goto out_release_pmic_glink;
+	}
+
 	return 0;
 
+out_release_pmic_glink:
+	mutex_lock(&__pmic_glink_lock);
+	__pmic_glink = NULL;
+	mutex_unlock(&__pmic_glink_lock);
 out_release_aux_devices:
 	if (pg->client_mask & BIT(PMIC_GLINK_CLIENT_BATT))
 		pmic_glink_del_aux_device(pg, &pg->ps_aux);
@@ -372,6 +382,8 @@ static void pmic_glink_remove(struct platform_device *pdev)
 	if (pg->client_mask & BIT(PMIC_GLINK_CLIENT_UCSI))
 		pmic_glink_del_aux_device(pg, &pg->ucsi_aux);
 
+	unregister_rpmsg_driver(&pmic_glink_rpmsg_driver);
+
 	guard(mutex)(&__pmic_glink_lock);
 	__pmic_glink = NULL;
 }
@@ -394,31 +406,7 @@ static struct platform_driver pmic_glink_driver = {
 		.of_match_table = pmic_glink_of_match,
 	},
 };
-
-static int pmic_glink_init(void)
-{
-	int ret;
-
-	ret = platform_driver_register(&pmic_glink_driver);
-	if (ret < 0)
-		return ret;
-
-	ret = register_rpmsg_driver(&pmic_glink_rpmsg_driver);
-	if (ret < 0) {
-		platform_driver_unregister(&pmic_glink_driver);
-		return ret;
-	}
-
-	return 0;
-}
-module_init(pmic_glink_init);
-
-static void pmic_glink_exit(void)
-{
-	unregister_rpmsg_driver(&pmic_glink_rpmsg_driver);
-	platform_driver_unregister(&pmic_glink_driver);
-}
-module_exit(pmic_glink_exit);
+module_platform_driver(pmic_glink_driver);
 
 MODULE_DESCRIPTION("Qualcomm PMIC GLINK driver");
 MODULE_LICENSE("GPL");
-- 
2.52.0

